packages/@aws-cdk/assertions/lib/assertions.ts

```diff
@@ -53,23 +54,49 @@ export class TemplateAssertions {
   /**
    * Assert that a resource of the given type and properties exists in the
    * CloudFormation template.
+   * By default, performs partial matching on the `Properties` key of the resource, via the
+   * `Match.objectLike()`. To configure different behavour, use other matchers in the `Match` class.
    * @param type the resource type; ex: `AWS::S3::Bucket`
    * @param props the 'Properties' section of the resource as should be expected in the template.
    */
   public hasResourceProperties(type: string, props: any): void {
-    const assertion = assert.haveResource(type, props, assert.ResourcePart.Properties);
-    assertion.assertOrThrow(this.inspector);
+    this.hasResource(type, Match.objectLike({
+      Properties: Match.isMatcher(props) ? props : Match.objectLike(props, { partial: true }),
+    }, { partial: true }));
   }
 
   /**
    * Assert that a resource of the given type and given definition exists in the
    * CloudFormation template.
+   * By default, performs partial matching on the resource, via the `Match.objectLike()`.
+   * To configure different behavour, use other matchers in the `Match` class.
    * @param type the resource type; ex: `AWS::S3::Bucket`
    * @param props the entire defintion of the resource as should be expected in the template.
    */
-  public hasResourceDefinition(type: string, props: any): void {
-    const assertion = assert.haveResource(type, props, assert.ResourcePart.CompleteDefinition);
-    assertion.assertOrThrow(this.inspector);
+  public hasResource(type: string, props: any): void {
+    const matcher = Match.isMatcher(props) ? props : Match.objectLike(props, { partial: true });
+    let closestResult: MatchFailure[] | undefined = undefined;
+    let count: number = 0;
+    for (const logicalId of Object.keys(this.inspector.value.Resources || {})) {
+      const resource = this.inspector.value.Resources[logicalId];
+      if (resource.Type === type) {
+        count++;
+        const result = matcher.test(resource);
+        if (result.length === 0) {
+          return;
+        }
+        if (closestResult === undefined || closestResult.length > result.length) {
+          closestResult = result;
+        }
+      }
+    }
+    if (closestResult === undefined) {
+      throw new Error(`No resource with type ${type} found`);
+    }
+    const messages: string[] = closestResult.map(r => `\t${r.message} at ${r.path.join('')}`);
+    messages.unshift(`${count} resources with type ${type} found, but none match as expected.` +
```

Comment:
Don't forget to print the data of the closest match as well.

Reply:
And it would be a good idea to print the matcher, too.

Ideally we'd print them together in some sort of diff view, that would be ace.

Reply:
> Don't forget to print the data of the closest match as well.

I am printing the output of the `closestResult`. Something else you mean?

Reply:
That prints the errors, not the object you were trying to match.

Reply:
Got it.